### ISSUE TITLE
segfault in Test_Filtering

### ISSUE THREAD

**Issue by kidder** [sxs-collaboration/spectre#3298]
# Bug reports:

### Expected behavior:

Test passes

### Current behavior:

Test consistently segfaults on my new desktop

### Environment:

Add as an attachment `$SPECTRE_BUILD_DIR/LibraryVersions.txt` or
add its contents here.
```
CMake Version: 3.20.3
SpECTRE Version: 2021.06.08
Git description: v2021.06.08-26-g63ddd991f
Git branch: develop
Git hash: 63ddd991fc2f18674a8e93b14057c691278d973f
Hostname: spectre2
Host system: Linux-5.12.8-300.fc34.x86_64
Host system version: 5.12.8-300.fc34.x86_64
Host system processor: x86_64
CMAKE_CXX_FLAGS: 
CMAKE_CXX_LINK_FLAGS:  -L/home/lek/ExternalPackages/charm/multicore-linux-x86_64/lib -fuse-ld=lld
CMAKE_CXX_FLAGS_DEBUG: -g -DSPECTRE_DEBUG
CMAKE_CXX_FLAGS_RELEASE: -O3 -DNDEBUG
CMAKE_C_FLAGS: 
CMAKE_C_FLAGS_DEBUG: -g
CMAKE_C_FLAGS_RELEASE: -O3 -DNDEBUG
CMAKE_Fortran_FLAGS: 
CMAKE_Fortran_FLAGS_DEBUG: -g
CMAKE_Fortran_FLAGS_RELEASE: -O3 -DNDEBUG -O3
CMAKE_EXE_LINKER_FLAGS: 
CMAKE_BUILD_TYPE: Debug
CMAKE_CXX_COMPILER: /usr/lib64/ccache/clang++
CMAKE_CXX_COMPILER_VERSION: 12.0.0
CMAKE_C_COMPILER: /usr/lib64/ccache/clang
CMAKE_C_COMPILER_VERSION: 12.0.0
CMAKE_Fortran_COMPILER: /usr/bin/f95
CMAKE_Fortran_COMPILER_VERSION: 11.1.1
Python Version: 3.9.5
```

### Detailed discussion:

Running `bin/RunTests Unit.Numerical.LinearOperators.Filter` through gdb, the segfault occurs at the lambda `cache_function` in the following code from `src/NumericalAlgorithms/LinearOperators/ExponentialFilter.cpp`

```
template <size_t FilterIndex>
const Matrix& Exponential<FilterIndex>::filter_matrix(const Mesh<1>& mesh) const
    noexcept {
  const auto cache_function = [this](
      const size_t extents, const Spectral::Basis basis,
      const Spectral::Quadrature quadrature) noexcept {
    return Spectral::filtering::exponential_filter(
        Mesh<1>{extents, basis, quadrature}, alpha_, half_power_);
  };

  const static auto cache = make_static_cache<
      CacheRange<1_st,
                 Spectral::maximum_number_of_points<Spectral::Basis::Legendre> +
                     1>,
      CacheEnumeration<Spectral::Basis, Spectral::Basis::Legendre,
                       Spectral::Basis::Chebyshev>,
      CacheEnumeration<Spectral::Quadrature, Spectral::Quadrature::Gauss,
                       Spectral::Quadrature::GaussLobatto>>(cache_function);
  return cache(mesh.extents(0), mesh.basis(0), mesh.quadrature(0));
}
```
with `extents = 1; basis = Spectral::Basis::Legendre; quadrature = Spectral::Quadrature::Gauss`.

In the test, if I comment out various combinations of the individual function calls in the following function

```
template <size_t Dim, bool FilterIndividually>
void invoke_test_exponential_filter_action(
    const double alpha, const unsigned half_power,
    const bool disable_for_debugging) noexcept {
  test_exponential_filter_action<Dim, Spectral::Basis::Legendre,
                                 Spectral::Quadrature::GaussLobatto,
                                 FilterIndividually>(alpha, half_power,
                                                     disable_for_debugging);
  test_exponential_filter_action<Dim, Spectral::Basis::Legendre,
                                 Spectral::Quadrature::Gauss,
                                 FilterIndividually>(alpha, half_power,
                                                     disable_for_debugging);
  test_exponential_filter_action<Dim, Spectral::Basis::Chebyshev,
                                 Spectral::Quadrature::GaussLobatto,
                                 FilterIndividually>(alpha, half_power,
                                                     disable_for_debugging);
  test_exponential_filter_action<Dim, Spectral::Basis::Chebyshev,
                                 Spectral::Quadrature::Gauss,
                                 FilterIndividually>(alpha, half_power,
                                                     disable_for_debugging);
}
```

it is always the second function call that triggers the segfault (with the  corresponding basis, quadrature, and lowest extent being passed to the lambda), and if there is only a single function call the test passes.

Putting a print statement at the start of the lambda, it appears the segfault occurs before the body of the lambda is entered.

Recall that the `StaticCache` is lazy, evaluating its functor on first call.

My current working theory is the segfault is caused by undefined behavior or a compiler bug caused by the lambda capturing `this` as `this` changes with each call of the  `test_exponential_filter_action` templated on Basis and Quadrature

I can get the tests to pass by declaring the lambda to be `const static auto`

But this leads me to following questions: 
- Why am I the first person to encounter this?  Am I the first to use clang 12.0.0?
- What is the expected behavior of a static StaticCache passed a lambda that captures `this` if `this` is destroyed? 
- It only makes sense for the `StaticCache` to use a lambda that captures if the objects captured are globally const, so wouldn't it make sense to capture static variables?
- Is declaring the lambda to be static fixing a bug, or changing undefined behavior (or bypassing a compiler bug) in such a way that the test no longer segfaults?
- The documentation for `Filters::Exponential` states that the template parameter `FilterIndex` is used to distinguish filters with different member variables (the ones used in the lambda), but as far as I can tell there is no check that all instances of each class are created with the same member variables.  Any ideas on how to such a check?

**Comment by kidder** [comment]
@moxcodes @nilsdeppe @wthrowe any thoughts or further things for me to try and diagnose the problem better are appreciated

**Comment by nilsdeppe** [comment]
- newest I've used is clang 11
- that code does look a bit questionable. Can you try capture `alpha_` and `half_power_` by value?

Re your last point: there's no way to check that as far as I'm aware.

**Comment by moxcodes** [comment]
I agree that avoiding the `[this]` capture sounds like a promising way to improve the problem

Regarding the check, would it work to define a static variable in the retrieval function, and then check that it is the same as the class member variable? I'm imagining something like:
```
static const double local_alpha = alpha_;
if (local_alpha != alpha_) {
  ERROR("uh-oh....");
}
```

My understanding of the static variables in a function is that they are initialized once according to the assignment when the function is first called, and the assignment is 'skipped' if the static variable already exists.

**Comment by nilsdeppe** [comment]
yea, I think that would work :)

Edit: comparing `double`s is always a bit dicey on a parallel system, so hopefully it works out

**Comment by wthrowe** [comment]
`cache_function` is being stored as a reference in the `StaticCache`, so it is a problem when it goes out of scope at the end of `filter_matrix`.  I don't see a good reason for that, so it is probably a bug in `StaticCache`.

The lambda should also be capturing everything it needs by value, as people have said above.

**Comment by kidder** [comment]
An update on this.  

-  Capturing the member variables by value or capturing `*this` (which captures a copy of the object) changed the behavior from a segfault to an error caused by the values of alpha and half power in the cached function being zero on the second call to `test_exponential_filter_action` that is templated on the Basis.  
- However, I have yet been able to make a simpler test that shows the behavior (e.g. instead of setting up a DataBox and using the Action testing framework, just  setting up the filters, everything works fine). 
- What does work, is moving the lambda directly into the call to make_static_cache (i.e. it is not named anymore).  This works for capturing `this`, `*this` or the member variables
- I will push that as a "fix" for now, but we may want to investigate more before closing the issue

**Comment by nilsdeppe** [comment]
@kidder did this get fixed?

**Comment by kidder** [comment]
"fixed" (by #3340) by inlining the lambda, but there may still be a bug in StaticCache
